### PR TITLE
Cap Size to [1-100] by clamping

### LINKED ISSUE #137: Size is not capped to min and max
There is nothing preventing `Size` from having a value over the `max` of `100` and therefore `percentage` returning a value higher than `1`.

Similarly the size can be negative which I don't think is valid (or is it?).

One confusing implication for this is when it comes to `Range`'s which scale with the size. I would expected `Range.linear(1, 1000)` to never produce bounds outside of these limits but it can if given a bad size. This leads to a generator producing unexpected values and the whole thing becomes a bit of a nightmare to try and track down. Of course the original bug is probably a bad `resize` or something but finding where that is is really tough since the error occurs when the generator is run not when it is constructed.

Is it intended for the `Size` to only ever be between `0` and `100`? If so should we cap it?

### PR DESCRIPTION
**Pull request by dwijnand**
Fixes #137

Feedback welcome.

### CODE REVIEW

**Comment by dwijnand** [comment]
Just noticed in `ProjectTReporting#report:

```scala
      if (size.value > Size.max)
        loop(successes, discards, sizeInit, seed, coverage)
```

Asides from dropping that for clean up purposes, will this change affect the running?  I saw the mention of #57, but I didn't get stuck into understanding it.

**Comment by charleso** [comment]
> Asides from dropping that for clean up purposes, will this change affect the running?

I doubt it. That's from the original Haskell version I'm pretty sure. You should remove it. There's also some size mode stuff a few lines up that you might be able to tidy slightly if you want.

Thanks Dale, I'm sorry size is so sloppy. :(

**Comment by dwijnand** [comment]
I'd initially opted for modulating because I saw it as the less lossy change to the user's input.  But the fact that I needed clamping in `report` lead me to switching.  LMK what you think.

File: core/shared/src/main/scala/hedgehog/core/PropertyT.scala [hunk 1]
```diff
@@ -168,14 +168,12 @@ trait PropertyTReporting {
 
   def report(config: PropertyConfig, size0: Option[Size], seed0: Seed, p: PropertyT[Result]): Report = {
     // Increase the size proportionally to the number of tests to ensure better coverage of the desired range
-    val sizeInc = Size(Math.max(1, Size.max / config.testLimit.value))
-    // Start the size at whatever remainder we have to ensure we run with "max" at least once
-    val sizeInit = Size(Size.max % Math.min(config.testLimit.value, Size.max)).incBy(sizeInc)
+    val sizeInc = Size(Size.max / config.testLimit.value)
+    // The increment value is also the starting value because size doesn't start at 0.
+    val sizeInit = sizeInc
```
Comment (dwijnand) [inline core/shared/src/main/scala/hedgehog/core/PropertyT.scala:173]: Double-check my logic (I checked a little in the REPL):

* `Math.min(config.testLimit.value, Size.max)` means 100+
* `Size.max % Math.min(config.testLimit.value, Size.max)` means `100 % 100+`, which is always 0
* (the `/ config.testLimit.value` above implies it's, at least, not 0 - could still be negative though...)
* `Size(0).incBy(sizeInc)` (before my value-correcting change here) meant simply `sizeInc`

**Comment by charleso** [comment]
Hey Dale. I'm really sorry but you had it right the first time. It's definitely meant to be modding and not clamping.

> Math.min(config.testLimit.value, Size.max) means 100+

Min not max. This was added to handle when testLimit was small and to make sure sizeInc would increase. If limit was 1 then sizeInc would be 100, limit 2 sizeInc 50, up to 100 at which point sizeInc would stay as 1.

Sorry if I derailed you, happy with your change as before, just thought there was a other mod we might be able to remove. No pressure to try and do more on this PR.

Thanks again for fixing my dodgy (untested) code.

**Comment by dwijnand** [comment]
Thanks for checking my logic - no matter how many checks and double-checks I still can't get my head around grokking min/max code! 😆

I tweaked that code to keep it producing the same sizeInit value as before, without modulus affecting it.  (Hopefully there isn't similar user-land code, famous last words...)

I think it's right now, but don't hesitate to point other things out.

**Comment by charleso** [comment]
@dwijnand Thanks for that!

**Pull request merged**